### PR TITLE
chore(sdk): Make the "error handler" give more information

### LINKED ISSUE #769: [TASK] Make the "error handler" give more information
Currently, the [`errorHandler`](https://github.com/Consensys/linea-attestation-registry/blob/f340f7c10bc6d4173e0c907eadcc21ce5b879091/sdk/src/utils/errorHandler.ts#L4) can only throw contextualized errors when their signature matches with the ABIs contained in the SDK.

1. In the case where the error is of type `ContractFunctionRevertedError`, but doesn't have an "errorName", use its signature.
2. In the case where the error is not of type `ContractFunctionRevertedError`, use its "shortMessage".
3. In the fallback case, change the generic message to `${type} failed`.

### PR DESCRIPTION
**Pull request by Chirag-S-Kotian**
## What does this PR do?

This PR refactors the error handling mechanism in the SDK to handle various error scenarios more effectively. It improves the clarity and specificity of error messages returned when a transaction fails, especially when dealing with contract function revert errors.

### Related ticket
Fixes #769

### Type of change
- [x] Chore
- [ ] Bug fix
- [ ] New feature
- [ ] Documentation update

## Check list
- [x] Unit tests for any smart contract change
- [ ] Contracts and functions are documented

---

### CODE REVIEW

File: sdk/src/utils/errorHandler.ts [hunk 1]
```diff
@@ -5,12 +5,15 @@ export function handleError(type: ActionType, err: unknown): never {
   if (err instanceof BaseError) {
     const revertError = err.walk((err) => err instanceof ContractFunctionRevertedError);
     if (revertError instanceof ContractFunctionRevertedError) {
-      const errorName = revertError.data?.errorName ?? "";
+      const errorName = revertError.data?.errorName ?? revertError.signature ?? "unknown revert reason";
```
Comment (alainncls) [inline sdk/src/utils/errorHandler.ts:8]: Let's avoid this nested ternary expression, it's quite hard to read/maintain.
Instead you could create a small `extractErrorName(revertError: ContractFunctionRevertedError)` function to make it easier to change the wording or content when throwing our error.

Also, in the default case, `errorName` should be "an unknown revert error" to make the full message perfect in the end.

File: sdk/src/utils/errorHandler.ts [hunk 2]
```diff
@@ -5,12 +5,15 @@ export function handleError(type: ActionType, err: unknown): never {
...
       throw new Error(`${type} failed with ${errorName}`);
+    } else {
+      const shortMessage = err.shortMessage ?? "An unknown error occurred";
```
Comment (alainncls) [inline sdk/src/utils/errorHandler.ts:11]:
```suggestion
      const shortMessage = err.shortMessage ?? "an unknown error";
```

To make the default error message smoother 🙂

**Review by alainncls — CHANGES_REQUESTED** [review]

**Comment by alainncls** [comment]
@Chirag-S-Kotian there seems to be an issue with the unit tests, I'll look at them once they're fixed, ok?
I've added 2 small comments on the implementation itself 🙂

**Comment by Chirag-S-Kotian** [comment]
Okay I will implement that and if there is an error in unit test please let me know.
Thank you 🙏😊.

File: sdk/src/utils/errorHandler.ts (hunk 2 above)
Comment (Chirag-S-Kotian) [reply sdk/src/utils/errorHandler.ts:11]: done.

**Comment by Chirag-S-Kotian** [comment]
hy @alainncls help me to run the tests locally using pnpm ,tell me the exact command to test the errorhandler.

**Comment by alainncls** [comment]
> hy @alainncls help me to run the tests locally using pnpm ,tell me the exact command to test the errorhandler.

If you want to run all unit tests for the SDK: `pnpm --filter "{sdk}" run test:unit` (from the root of the project)
And if you want to run only the `errorHandler` unit tests: `pnpm --filter "{sdk}" run test:unit -- src/utils/errorHandler.test.ts`

Hope that helps!

**Review by Nmhth17 — APPROVED** [review]

**Comment by Chirag-S-Kotian** [comment]
@Nmhth17 review 😌 please .

File: sdk/src/utils/errorHandler.ts [hunk 3]
```diff
@@ -1,16 +1,29 @@
 import { BaseError, ContractFunctionRevertedError } from "viem";
 import { ActionType } from "./constants";
 
+function extractErrorName(revertError: ContractFunctionRevertedError): string {
+  if (revertError.data?.errorName) {
+    return revertError.data.errorName;
```
[the remark this answers is not in the document]
Comment (Chirag-S-Kotian) [reply sdk/src/utils/errorHandler.ts:6]: i resolved it .

**Review by alainncls — APPROVED** [review]

**Pull request merged**